src/api/block_builder/input_selection/helpers.rs

```diff
@@ -0,0 +1,257 @@
+
+        if let Some(output_native_tokens) = output.native_tokens() {
+            required_native_tokens.add_native_tokens(output_native_tokens.clone())?;
+        }
+    }
+
+    // check if a foundry mints or melts native tokens
+    let (minted_native_tokens, melted_native_tokens) = get_minted_and_melted_native_tokens(inputs, outputs)?;
+    // add melted native tokens as outputs, because we need to have this amount in the inputs
+    required_native_tokens.merge(melted_native_tokens)?;
+
+    Ok(AccumulatedOutputAmounts {
+        minted_native_tokens,
+        amount: required_amount,
+        native_tokens: required_native_tokens,
+    })
+}
+
+/// Computes the minimum storage deposit amount that a basic output needs to have with an [AddressUnlockCondition] and
+/// optional [NativeTokens].
+pub fn minimum_storage_deposit_basic_output(
+    config: &RentStructure,
+    address: &Address,
+    native_tokens: &Option<NativeTokens>,
+) -> Result<u64> {
+    let address_condition = UnlockCondition::Address(AddressUnlockCondition::new(*address));
+    let mut basic_output_builder = BasicOutputBuilder::new_with_amount(OutputAmount::MIN)?;
+    if let Some(native_tokens) = native_tokens {
+        basic_output_builder = basic_output_builder.with_native_tokens(native_tokens.clone());
+    }
+    let basic_output = basic_output_builder
+        .add_unlock_condition(address_condition)
+        .finish_output()?;
+
+    Ok(basic_output.rent_cost(config))
+}
+
+/// Get the `StorageDepositReturnUnlockCondition`, if not expired
+pub(crate) fn sdr_not_expired(output: &Output, current_time: u32) -> Option<&StorageDepositReturnUnlockCondition> {
+    if let Some(unlock_conditions) = output.unlock_conditions() {
+        if let Some(sdr) = unlock_conditions.storage_deposit_return() {
+            let expired = if let Some(expiration) = unlock_conditions.expiration() {
+                current_time >= expiration.timestamp()
+            } else {
+                false
+            };
+
+            // We only have to send the storage deposit return back if the output is not expired
+            if !expired { Some(sdr) } else { None }
+        } else {
+            None
+        }
+    } else {
+        None
+    }
+}
+
+// Inputs need to be sorted before signing, because the reference unlock conditions can only reference a lower index
+pub(crate) fn sort_input_signing_data(inputs: Vec<InputSigningData>) -> crate::Result<Vec<InputSigningData>> {
+    // filter for ed25519 address first, safe to unwrap since we encoded it before
+    let mut sorted_inputs = inputs
+        .clone()
+        .into_iter()
+        // PANIC: safe to unwrap as we encoded the address before
+        .filter(|input| Address::try_from_bech32(&input.bech32_address).unwrap().1.kind() == Ed25519Address::KIND)
+        .collect::<Vec<InputSigningData>>();
+
+    for input in &inputs {
+        // Don't add outputs duplicated
+        if sorted_inputs.iter().any(|i| {
```

Comment:
This is a N^2 complexity so not idea. If I understand it well, you want to insert what is not Ed25519Address to the right position so instead of `filter` L96 you could use `partition` and then iterate over the rest of the inputs, then completely removing this if.